// wizzi-core/tests/production/mocks/schemaLoaders.js
var verify = require('../../../lib/util/verify');

var md = module.exports = {};

md.xml = function (filepath, context, callback) {
    if (verify.endsWith(filepath, 'blue.xml.ittf')) {
        callback(null, xml_blue(context));
    } else if (verify.endsWith(filepath, 'red.xml.ittf')) {
        callback(null, xml_red(context));
    } else if (verify.endsWith(filepath, 'green.xml.ittf')) {
        callback(null, xml_green(context));
    } else {
        throw new Error("Cannot load xml wmt model for filepath : " + filepath);
    }
}

md.rdbms = function (filepath, context, callback) {
    if (verify.endsWith(filepath, 'test.rdbms.ittf')) {
        callback(null, rdbms_test(context));
    } else {
        throw new Error("Cannot load xml wmt model for filepath : " + filepath);
    }
}

function rdbms_test(context) {
    var ret = {
        WmtName: 'rdbms',
        tables: []
    };
    ret.tables.push({
        WmtName: 'User',
        columns: [
            { WmtName: 'email' }
        ]
    });
    ret.tables.push({
        WmtName: 'UserPref',
        columns: [
            { WmtName: 'key' },
            { WmtName: 'value' }
        ]
    });
    return ret;
}
function xml_blue(context) {
    var ret = {
        name: 'rdbms',
        childs: []
    };
    context.rdbms.tables.forEach(function (item) {
        ret.childs.push({
            name: 'bluetable', value: item.WmtName, childs: [] });
    });
    return ret;
}
function xml_red(context) {
    var ret = {
        name: 'rdbms',
        childs: []        
    };
    context.rdbms.tables.forEach(function (item) {
        ret.childs.push({ name: 'redtable', value: item.WmtName, childs: [] });
    });
    return ret;
}
function xml_green(context) {
    var ret = {
        name: 'rdbms',
        childs: []
    };
    context.rdbms.tables.forEach(function (item) {
        ret.childs.push({ name: 'greentable', value: item.WmtName, childs: [] });
    });
    return ret;
}
